### ISSUE TITLE
ManyToMany relationship updates only one side with CRUD operations

### ISSUE THREAD

**Issue by nagwan** [vuex-orm/vuex-orm#303]
i am trying to work with many to many relationship in a complete CRUD system
the system is all about users each has an account or more, where the same account may be shared among group of users.

here in the models
```javascript
// user model
export default class User extends Model {
    static entity = "users"

    static fields(){
        return {
            id: this.increment(),
            first_name: this.string(),
            last_name: this.string(),
            email: this.string(),
            accounts: this.belongsToMany(Account, AccountUser, "user_id", "account_id")
        }
    }

 get full_name () {
        return `${this.first_name} ${this.last_name}`
    }
}

// account model

export default class Account extends Model {

    static entity = "accounts"

    static fields(){
        return {
            id: this.increment(),
            name: this.string(),
            accountInfo: this.string(),
            users: this.belongsToMany(User, AccountUser, "account_id", "user_id")
        }
    }
}

// the pivot model

export default class AccountUser extends Model {
    static entity = "account-users"

    static primaryKey = ["user_id", "account_id"]

    static fields() {
        return {
            user_id: this.attr(),
            account_id: this.attr(),
        }
    }
}

```
```javascript

// vuex store action to fetch data

actions: {
    fetchAccounts(context) {
        let accounts = [
            {
                id: 1,
                name: "The United Bank ",
                accountInfo: "Lorem ipsum dolor sit amet consectetur",
                users: [2]
            },
            {
                id: 2,
                name: "Vodafone cash ",
               accountInfo: "Lorem ipsum dolor sit amet consectetur",
            },

        ]

        Account.insert({ data: accounts })

    },
    fetchUsers(context) {
        let users = [
            {
                id: 1,
                first_name: "Adam",
                last_name: "Albert",
                email: "adam_albert@gmail.com",

            },
            {
                id: 2,
                first_name: "John",
                last_name: "Tom",
                email: "john_tom@gmail.com",
                accounts: [1]
            },

        ]

        User.insert({ data: users })
    },
}
```

```javascript

//display-accounts.vue

// in the actual code all the actions are dispatched in the main layout
beforeCreate(){
 this.$store.dispatch("fetchAccounts");
 this.$store.dispatch("fetchUsers");
}

 computed: {
    fetchAccounts: {
      get() {
        return Account.query().with("users").all();
      },
      set() {}
    },
  },

 methods: {
    editAccount(account) {
      this.$router.push({ name: "edit-account", params: { id: account } });
    },
 }

// component template (display data component)
  <div class>
    <div class="card" v-for="account in fetchAccounts" :key="account.id">
      <div class="card-body">
              <p >
                <b>Name:</b>
                {{account.name}}
              </p>
        
              <p >
                <b>Users:</b>
                <span v-if="account.users.length">{{account.users.length}}</span>
                <span v-else>None</span>
              </p>
      
        <div>
         <i class="fas fa-edit" @click="editAccount(account.id)"></i>
        </div>

      </div>
    </div>
  </div>

```
if we try to update an account everything works just perfect and all the fields updated except for the users field, the changes are discarded regardless we added or removed users

```javascript

// new-account-form.vue

<script>
export default {
  data() {
    return {
      // (Form) is class i am using to handle forms and errors
      newAccount: new Form({
        name: null,
        accountInfo: null,
        users: []
      }),
      accountToEditId: null
    };
  },

  mounted() {
    this.fetchData();
  },

  watch: {
    $route: {
      deep: true,
      handler() {
        this.newAccount = this.fetchData();
      }
    }
  },

  computed: {
    fetchUsers: {
      get() {
        return User.query().all();
      },
      set() {}
    }
  },

  methods: {
    updateAccount() {
      Account.update({
        where: this.accountToEditId,
        data: this.newAccount
      });
      return this.$router.push({ name: "display-accounts" });
    },

    fetchData() {
      if (this.$route.params.id === undefined) {
        return this.newAccount;
      } else {
        let id = parseInt(this.$route.params.id);

        this.accountToEditId = id;

        let accountToEdit = _.cloneDeep(
          Account.query()
            .whereId(id)
            .withAll()
            .first()
        );

     // extracting users idz in lieu of the whole user object
        let getUsersIdz = _.map(accountToEdit.users, user => {
          return user.id;
        });

        return (this.newAccount = new Form({
          name: accountToEdit.name,
          accountInfo: accountToEdit.accountInfo,
          users: getUsersIdz
        }));
      }
    },
  cancel() {
      this.$router.push({ name: "display-accounts" });
    },
  }
};
</script>


<template>
  <div class="d-flex align-items-center justify-content-center">
    <form
      class="w-100"
      @submit.prevent="updateAccount"
      method="POST"
    >
      <input
        class="form-control"
        type="text"
        v-model="newAccount.name"
        placeholder="Account Name (EN)"
      >
      <textarea class="form-control" v-model="newAccount.accountInfo" placeholder="Account Info"></textarea>

      <!-- users -->
      <div class="row justify-content-center">
        <div
          class="col-lg-3 col-md-4 col-6 m-t-30 m-b-30"
          v-for="user in fetchStaff"
          :key="user.id"
        >
          <div class="d-flex justify-content-center align-items-center m-b-5">
            <div class="custom-control card-checkbox">
              <input
                type="checkbox"
                :value="user.id"
                class="custom-control-input"
                v-model="newAccount.users"
                :id="'user_'+user.id"
              >
              
              <label class="custom-control-label" :for="'user_'+user.id">
                <h5 class>
                  <b>Name:</b>
                  {{user.full_name}}
                </h5>
              </label>

            </div>
          </div>
        </div>
      </div>

      <div class="row align-items-center justify-content-center m-t-30 m-b-30">
       
        <div class="col-md-2 col-6" v-if="accountToEditId">
          <button type="submit" class="btn btn-big btn-success w-100">Update</button>
        </div>

        <div class="col-md-2 col-6">
          <button type="button" class="btn btn-big btn-light w-100" @click="cancel">Cancel</button>
        </div>
      </div>
    </form>
  </div>
</template>

```

in case of adding a new account and attach users to it, the length of the accounts array for the attached users will change,but the users array length for this new account will be 0.

and in the pivot model, a new record inserted with the  selected user_id but with a null account_id 

(and in {display-users component} all the account data defaults to null except the id) 

and it goes the same if we track the users 

sorry for long description but trying to make it as descriptive as i can

**Comment by kiaking** [comment]
@nagwan Hi! thank you for the report.

I think you're not using `fetchUsers` defined as `computed` property in your template? For change to be visible – reactive – you need to use that computed property to display records.

**Comment by nagwan** [comment]
@kiaking  if u mean in display-users component, no i am following the same pattern in the users side, i just found the issue became so long so i wrote only the accounts side

in display users-component i am using the following computed to fetch users
```
 computed: {
    fetchUsers: {
      get() {
        return User.query().with("accounts").all();
      },
      set() {}
    },
  },
```

thank u for your efforts and time

**Comment by kiaking** [comment]
@nagwan Oh, OK. Thank you for the explanation. I need to take a bit time to look into this. I'll get back to you if I see any progress!

**Comment by nagwan** [comment]
and btw if i am not using computed no changes will appear in the users side, but what i have here is different the account is added to the selected users but the users don`t added to the account 

thank u waiting for your replay

**Comment by Gilux** [comment]
If it can help, I think I ran into a similar problem :

I have Company, CompanyTechno and Techno models
```js
class Company extends Model {
  // This is the name used as module name of the Vuex Store.
  static entity = "company";

  // List of all fields (schema) of the post model. `this.attr` is used
  // for the generic field type. The argument is the default value.
  static fields() {
    return {
      id: this.increment(),
      company_name: this.attr(""),
      // ...
      technos: this.belongsToMany(Techno, CompanyTechno, "company_id", "techno_id" )
    };
  }
}

// Pivot class
class CompanyTechno extends Model {
  static entity = 'companyTechno'

  static primaryKey = ['company_id', 'techno_id']

  static fields () {
    return {
      company_id: this.attr(null),
      techno_id: this.attr(null)
    }
  }
}

// Techno class
class Techno extends Model {
  // This is the name used as module name of the Vuex Store.
  static entity = 'techno'

  // List of all fields (schema) of the post model. `this.attr` is used
  // for the generic field type. The argument is the default value.
  static fields () {
    return {
      id: this.increment(),
      name: this.string(""),
      logo_url: this.string(""),
      companies: this.belongsToMany(Company, CompanyTechno, 'techno_id', 'company_id')
    }
  }
}
```

In my AddCompany component which is a form, I have a multiple select which gives me an array of techno IDs.

After saving my new `Company` entity, I have 2 `CompanyTechno` entities created, each one with `techno_id` filled, but `company_id` is null.

As a temporary fix, I replaced `id: this.increment(),` with `id: this.number(null),` in my Company model class, and when I submit the form, `id` is filled with `Date.now()`.

**Comment by nagwan** [comment]
@Gilux , it looks good but what is the use of `Date now()`
  as a temporary solution also i tried something different, i used promises to get the new added account id then i used  `` User.query().whereIdIn(newAccount.users).get()``
then i startd to map over each user and add the new account id to  the accounts array associated to each user , and i dropped the relationship 😞

**Comment by Gilux** [comment]
@nagwan I use `Date.now()` to get the current timestamp (in milliseconds). In my use case, it's impossible that a user may create 2 `Company` entities at the same time, so I'm sure it will be unique and I won't encounter duplicate primary key issues.
But it remains certainly a dirty solution 😆

**Comment by nagwan** [comment]
so we will have to wait till vuex orm gives us a clean solution 🙃

**Comment by pimeo** [comment]
@kiaking After made some researches few days ago, I think the main trouble is about using the increment method in Model class that does not work correctly when it is coupled to a primaryKey and used as a relation (and also whenever the primaryKey property name use to match any relations).

Relationship management works if the property name used as primaryKey is pre-filled before inserting record in a model.

**Comment by nagwan** [comment]
@pimeo 

> Relationship management works if the property name used as primaryKey is pre-filled before inserting record in a model.<

i don`t understand this part

**Comment by kiaking** [comment]
OK, I've created test case and I think I've reproduced this. It happens when combining with `increment` attribute. I'll be working on to this.

**Comment by kiaking** [comment]
Finally this is fixed with Vuex ORM 0.31.11! Thank you for the report and the patients!  
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.31.11
